File: ArrayAndStrins/c1e3.cpp
#include <iostream>

using namespace std;

string urlify(char str[])
{
	int i;
	int spaceCount = 0;
	
	for (i = 0; str[i] ;  i++)
	{
		if (str[i] == ' ') spaceCount++;
	}

	while (str[i-1] == ' ')
	{
		spaceCount--;
		i--;
	}

	int length = i + 2 * spaceCount;
	int index = length - 1;

	str[index--] = '\0';
	
	for (int j = i-1; j >= 0; j--)
	{
		if (str[j] == ' ')
		{
			str[index] = '0';
			str[index-1] = '2';
			str[index-2] = '%';

			index -= 3;
		}
		else
		{
			str[index] = str[j];
			index--;
		}
	}
	cout << length;
	return str;
}

int main(int argc, char* argv[])
{
	for (int i = 1; i < argc; i++)
	{
		cout << argv[i] << " : " << urlify(argv[i]) << endl; 
	}

	return 0;
}

File: ArrayAndStrins/c1e3.py
import array

def urlify(st: str) -> str:
    index = 0
    spaceCount = 0
    lastCharIndex = False

    for i in range (len(st) -1, -1, -1):
        if st[i] != ' ' and not lastCharIndex:
            lastCharIndex = i
        if lastCharIndex and st[i] == ' ': spaceCount+=1

    index = lastCharIndex + 2 * spaceCount
    
    st = array.array('u', st[:index+1])

    for i in range(lastCharIndex, -1, -1 ):
        if st[i] == ' ':
            st[index]='0';
            st[index-1]='2';
            st[index-2]='%';
            
            index-=3
        else:
            st[index] = st[i]
            index-=1
    
    return st.tounicode()

def urlifyPythonicWay(st: str) -> str:
    st = st.rstrip()
    return st.replace(' ', '%20')
